server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,296 @@
+     * @param remoteDirectory corresponding to the remote directory
+     * @param fileCache used to cache the remote files locally
+     */
+    public CompositeDirectory(FSDirectory localDirectory, RemoteSegmentStoreDirectory remoteDirectory, FileCache fileCache) {
+        super(localDirectory);
+        this.localDirectory = localDirectory;
+        this.remoteDirectory = remoteDirectory;
+        this.fileCache = fileCache;
+        transferManager = new TransferManager(
+            (name, position, length) -> new InputStreamIndexInput(
+                remoteDirectory.openInput(name, new BlockIOContext(IOContext.DEFAULT, position, length)),
+                length
+            ),
+            fileCache
+        );
+    }
+
+    /**
+     * Returns names of all files stored in this directory in sorted order
+     * Does not include locally stored block files (having _block_ in their names)
+     *
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public String[] listAll() throws IOException {
+        logger.trace("listAll() called");
+        String[] localFiles = localDirectory.listAll();
+        logger.trace("Local Directory files : {}", () -> Arrays.toString(localFiles));
+        Set<String> allFiles = new HashSet<>(Arrays.asList(localFiles));
+        String[] remoteFiles = getRemoteFiles();
+        allFiles.addAll(Arrays.asList(remoteFiles));
+        logger.trace("Remote Directory files : {}", () -> Arrays.toString(remoteFiles));
+        Set<String> localLuceneFiles = allFiles.stream()
+            .filter(file -> !FileType.isBlockFile(file))
+            .collect(Collectors.toUnmodifiableSet());
+        String[] files = new String[localLuceneFiles.size()];
+        localLuceneFiles.toArray(files);
+        Arrays.sort(files);
+        logger.trace("listAll() returns : {}", () -> Arrays.toString(files));
+        return files;
+    }
+
+    /**
+     * Removes an existing file in the directory.
+     * Currently deleting only from local directory as files from remote should not be deleted due to availability reasons
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void deleteFile(String name) throws IOException {
+        logger.trace("deleteFile() called {}", name);
+        if (isTempFile(name)) {
+            localDirectory.deleteFile(name);
+        } else {
+            /*
+            Not deleting from localDirectory directly since it causes a race condition when the localDirectory deletes a file, and it ends up in pendingDeletion state.
+            Meanwhile, fileCache on removal deletes the file directly via the Files class and later when the directory tries to delete the files pending for deletion (which happens before creating a new file), it causes NoSuchFileException and new file creation fails
+            */
+            fileCache.remove(localDirectory.getDirectory().resolve(name));
```

Comment:
Similar to `close` will `deleteFile` be called only when there is no reference to the `IndexInput` for the file being deleted ?

Reply:
+1, who all are the callers of deleteFile?

Reply:
Yes from our tests, we can see that deleteFile is being called only when there is no active ref on the file